src/crypto/openssl/symmetric_key.cpp

```diff
@@ -54,16 +54,39 @@ namespace ccf::crypto
     std::vector<uint8_t>& cipher,
     uint8_t tag[GCM_SIZE_TAG]) const
   {
-    std::vector<uint8_t> cb(plain.size());
-    int len = 0;
+    if (aad.empty() && plain.empty())
+    {
+      throw std::logic_error("aad and plain cannot both be empty");
+    }
+
     Unique_EVP_CIPHER_CTX ctx;
     CHECK1(EVP_EncryptInit_ex(ctx, evp_cipher, NULL, key.data(), NULL));
+
     CHECK1(EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, iv.size(), NULL));
     CHECK1(EVP_EncryptInit_ex(ctx, NULL, NULL, key.data(), iv.data()));
+
     if (!aad.empty())
-      CHECK1(EVP_EncryptUpdate(ctx, NULL, &len, aad.data(), aad.size()));
-    CHECK1(EVP_EncryptUpdate(ctx, cb.data(), &len, plain.data(), plain.size()));
-    CHECK1(EVP_EncryptFinal_ex(ctx, cb.data() + len, &len));
+    {
+      int aad_outl{0};
+      CHECK1(EVP_EncryptUpdate(ctx, NULL, &aad_outl, aad.data(), aad.size()));
+    }
+
+    std::vector<uint8_t> cb(plain.size());
+    if (!plain.empty())
+    {
+      int plain_outl{0};
+      CHECK1(EVP_EncryptUpdate(
+        ctx, cb.data(), &plain_outl, plain.data(), plain.size()));
+      assert(static_cast<size_t>(plain_outl) == plain.size());
+    }
+
+    int final_outl{0};
+    CHECK1(EVP_EncryptFinal_ex(ctx, NULL, &final_outl));
+
+    // Final outl must be 0, as we use no padding and we call EncryptFinal_ex to
+    // follow the common workflow and perform final checks if any.
```

Comment:
Can we state where specifically we decide not to use padding? Is it line 65?

Reply:
there is no padding with GCM, it's a consequence of the algorithm choice

Reply:
Thank you for the confirmation.

Could you probably point to any evidence of that? I'd like some link in the code for posterity, but I can't find it [here](https://docs.openssl.org/1.0.2/man3/EVP_EncryptInit/#description).

Reply:
Let's try to link to the 3.3 doc if we can https://docs.openssl.org/3.3/man3/EVP_EncryptInit/#flags
I think a reference to line 63, where we set the cipher to AES GCM (https://docs.openssl.org/3.3/man3/EVP_aes_128_gcm/) is probably a good idea.

Reply:
But this doesn't link the cipher mode to the padding. What I'm looking for is sentence in the doc saying "AES GCM uses no padding, block size = 1, so encypt_final always returns 0", or something similar.

Reply:
Ok, this seems to say it 
https://docs.openssl.org/3.3/man3/EVP_EncryptInit/#aead-interface